### PR TITLE
JDK-8248863: Add search landing page to API documentation

### PR DESCRIPTION
**Pull request by hns**
This is the second of two PRs to enhance JavaDoc search, it is based on the first one (#8185).

It adds a standalone search page (search.html) along with its own script file (search-page.js). This PR is very similar to the last prototype I uploaded and demoed, the changes are mostly tweaks to the markup, style sheets and text. 

JDK API docs rendered with this patch can be viewed and tested here (top-level files only, updated on April 20th):
http://cr.openjdk.java.net/~hannesw/8248863/api.02/
http://cr.openjdk.java.net/~hannesw/8248863/api.02/search.html

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 reviews required, with at least 1 reviewer)

### Issue
 * [JDK-8248863](https://bugs.openjdk.java.net/browse/JDK-8248863): Add search landing page to API documentation

### Reviewers
 * [Jonathan Gibbons](https://openjdk.java.net/census#jjg) (@jonathan-gibbons - **Reviewer**) ⚠️ Review applies to [27e057fb](https://git.openjdk.java.net/jdk/pull/8226/files/27e057fbc7d1147ac8820d6016e4a06e04c0122c)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8226/head:pull/8226` \
`$ git checkout pull/8226`

Update a local copy of the PR: \
`$ git checkout pull/8226` \
`$ git pull https://git.openjdk.java.net/jdk pull/8226/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8226`

View PR using the GUI difftool: \
`$ git pr show -t 8226`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8226.diff">https://git.openjdk.java.net/jdk/pull/8226.diff</a>

</details>

### CODE REVIEW

**Comment by liach** [comment]
User remark:
1. Seems the result list will automatically scroll down when the mouse hovers over, which feels weird.
2. When an arbitrary string is typed in and enter is pressed, it navigates to the best match, even though it may not be an accurate match to the class simple name or method name. Should we consider navigating to the results page first in that case?

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/resources/search-page.js [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+        setSearchUrl();
+        var term = searchTerm = input.val().trim();
+        if (term === "") {
+            notify.html(messages.enterTerm);
+            activeTab = "";
+            fixedTab = false;
+            resultContainer.empty();
+            resultSection.hide();
+        } else {
+            notify.html("Searching...");
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/resources/search-page.js:241]: hmmm, should this be localized?

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/resources/stylesheet.css [hunk 2]
```diff
@@ -644,6 +670,98 @@ ul.ui-autocomplete li {
...
+div#result-container a.search-result-link {
+    padding: 0;
+    margin: 4px 0;
+    width: 100%;
+}
+#result-container .result-highlight {
+    font-weight:bolder;
+}
+.page-search-info {
+    background-color: #f5f8ff;
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/resources/stylesheet.css:688]: reminder: we were going to look at CSS variables to better manage colors

**Comment by hns** [comment]
> User remark:
> 
>     1. Seems the result list will automatically scroll down when the mouse hovers over, which feels weird.

I think what you're referring to is a behavior of the jQuery autocomplete component we are using. It was not modified in this PR, however. 
 
>     2. When an arbitrary string is typed in and enter is pressed, it navigates to the best match, even though it may not be an accurate match to the class simple name or method name. Should we consider navigating to the results page first in that case?

I don't understand that. Can you provide an example please?

**Comment by hns** [comment]
@jonathan-gibbons thanks for the comments, I added a commit addressing the issues you pointed out.

**Comment by hns** [comment]
> > ```
> > 1. Seems the result list will automatically scroll down when the mouse hovers over, which feels weird.
> > ```
> 
> I think what you're referring to is a behavior of the jQuery autocomplete component we are using. It was not modified in this PR, however.

I see what you mean, and the scrolling behaviour has actually changed. I'll look into it.

**Comment by hns** [comment]
>     1. Seems the result list will automatically scroll down when the mouse hovers over, which feels weird.
> 
>     2. When an arbitrary string is typed in and enter is pressed, it navigates to the best match, even though it may not be an accurate match to the class simple name or method name. Should we consider navigating to the results page first in that case?

@liach I updated the pull request to fix the first problem. The main PR description links to the updated docs. I still don't understand the second issue. Thanks!

**Comment by liach** [comment]
Looks good now.
For 2, I was thinking about if we have ambiguous terms for input, like `abstract`, `exact` where there are multiple best-match, we might want to go to landing page than to the top result when pressing enter. But that's optional.

In addition, I didn't notice the **SEARCH** text that led to the search page at first. Since it leads to the search page itself, would it make sense to merged the "Go to search page" with that text (which does not preserve current text in search bar) when there's something typed in the search bar?

**Review by jonathan-gibbons — APPROVED** [review]
OK.

For the record, I note @hns gave us a walkthrough of the new JavaScript code,. One general result from that was a desire for more comments ...  at a high-level for a description of the general search algorithm, at a medium level for some of the more interesting functions and occasionally down at the detail level, when the code is not obviously self-documenting, such as in the logic to build the regular expressions for the search.  It would also help to note somewhere that the search is on a very specific dataset (the search index files generated by the doclet) and that we should document the format of those files somewhere. Finally, we discussed the idea to "partially minify" the JavaScript files, to strip out comments and unnecessary whitespace, so that there is no inhibition of the amount of comments that might be added and to reduce the download cost/time for end users. This processing could maybe be done in the makefiles at build time.

Comments and minifying the file can be done in a followup PR.

Great work.

**Pull request closed without merging**